src/core/internal/lifetime.d

```diff
@@ -84,23 +84,87 @@ if (is(UT == core.internal.traits.Unqual!UT))
     emplaceRef!(UT, UT)(chunk, forward!args);
 }
 
-//emplace helper functions
-private nothrow pure @trusted
-void emplaceInitializer(T)(scope ref T chunk)
+/+
+Emplaces T.init.
+In contrast to `emplaceRef(chunk)`, there are no checks for disabled default
+constructors etc.
++/
+void emplaceInitializer(T)(scope ref T chunk) nothrow pure @trusted
+    if (!is(T == const) && !is(T == immutable) && !is(T == inout))
 {
-    // Emplace T.init.
-    // Previously, an immutable static and memcpy were used to hold an initializer.
-    // With improved unions, this is no longer needed.
-    union UntypedInit
+    import core.internal.traits : hasElaborateAssign;
+
+    static if (!hasElaborateAssign!T && __traits(compiles, chunk = T.init))
     {
-        T dummy;
+        chunk = T.init;
     }
-    static struct UntypedStorage
+    else
+    {
+        static if (__traits(isZeroInit, T))
+        {
+            static if (is(T U == shared U))
+                alias Unshared = U;
+            else
+                alias Unshared = T;
+
+            import core.stdc.string : memset;
+            memset(cast(Unshared*) &chunk, 0, T.sizeof);
```

Comment:
Does this need to cast to the explicit type (as `memset` takes a `void*` anyway)?

```suggestion
            memset(cast(void*) &chunk, 0, T.sizeof);
```

Reply:
It doesn't need to, but I consider it good practice. As soon as someone has the great idea to drop the template constraints, a cast to `void*` will allow non-mutable types too.